compiler/noirc_evaluator/src/ssa/opt/brillig_entry_points.rs

```diff
@@ -80,40 +80,92 @@ impl Ssa {
             return self;
         }
 
+        // Build a call graph based upon the Brillig entry points and set up
+        // the functions needing specialization before performing the actual call site rewrites.
         let brillig_entry_points = get_brillig_entry_points(&self.functions, self.main_id);
-
         let functions_to_clone_map = build_functions_to_clone(&brillig_entry_points);
+        let (calls_to_update, mut new_functions_map) =
+            build_calls_to_update(&mut self, functions_to_clone_map, &brillig_entry_points);
 
-        let calls_to_update = build_calls_to_update(&mut self, functions_to_clone_map);
-
-        let mut new_functions_map = HashMap::default();
-        for (entry_point, inner_calls) in brillig_entry_points {
-            let new_entry_point =
-                new_functions_map.get(&entry_point).copied().unwrap_or(entry_point);
-
-            let function =
-                self.functions.get_mut(&new_entry_point).expect("ICE: Function does not exist");
-            update_function_calls(function, entry_point, &mut new_functions_map, &calls_to_update);
+        // Now we want to actually rewrite the appropriate call sites
+        // First pass to rewrite the originally supplied call graph
+        for CallToUpdate {
+            entry_point,
+            function_to_update,
+            instruction,
+            new_func_to_call: new_id,
+            call_args: arguments,
+        } in calls_to_update
+        {
+            let new_function_to_update = if entry_point == function_to_update {
+                entry_point
+            } else {
+                new_functions_map
+                    .entry(entry_point)
+                    .or_default()
+                    .get(&function_to_update)
+                    .copied()
+                    .unwrap_or(function_to_update)
```

Comment:
Is there any issue if a function is processed twice? I would expect to skip cases when no `get(&function_to_update)` is found

Reply:
`new_functions_map` only maps new old function call sites to new ones. However, we can still potentially still have an old function we need to update. The function id we are fetching here is for the caller, not the callee. Any new callee's that then need to be updated will be resolved in the follow-up pass.

Reply:
Also due to our mapping of old id -> new id per entry point, it should be fine to process a function twice as we will have the same result.